File: frontend/src/Objects/Tool.js
import Position from './Position';
import { createImage, copyImage } from '../Components/Tools.jsx';

class Tool {
	constructor(
		type,
		image,
		position = null,
		width,
		height,
		layer = null,
		color = '#0077be',
		amount = 0,
		taper = 0.5,
		capacity = 2,
		intervals = 4
	) {
		this.name = 'Tool ' + layer;
		this.type = type;
		this.image = image;
		this.position = position;
		this.width = width;
		this.height = height;
		this.layer = layer;
		this.color = color;
		this.amount = amount;
		this.taper = taper;
		this.capacity = capacity;
		this.intervals = intervals;
		this.new = true;
	}

	getName() {
		return this.name;
	}

	getType() {
		return this.type;
	}

	getImage() {
		return this.image;
	}

	getPosition() {
		return this.position;
	}

	getWidth() {
		return this.width;
	}

	getHeight() {
		return this.height;
	}

	getLayer() {
		return this.layer;
	}

	setName(name) {
		this.name = name;
	}

	setPosition(position) {
		this.position = position;
	}

	setPosition(x, y) {
		this.position = new Position(x, y);
	}

	setLayer(layer) {
		this.layer = layer;
	}

	getAmount() {
		return this.amount;
	}

	setAmount(amount) {
		this.amount = amount;
	}

	getColor() {
		return this.color;
	}

	setColor(color) {
		this.color = color;
	}

	setWidth(width) {
		this.width = width;
	}

	setHeight(height) {
		this.height = height;
	}

	setLayer(layer) {
		this.layer = layer;
	}

	isNew() {
		return this.new;
	}

	stale() {
		this.new = false;
	}

	equals(tool) {
		return (
			this.name == tool.name &&
			this.type == tool.type &&
			this.position.x == tool.position.x &&
			this.position.y == tool.position.y &&
			this.layer == tool.layer &&
			this.imageEquals(tool)
		);
	}

	imageEquals(tool) {
		const { properties, animation } = this.image;
		for (const property in properties) {
			if (properties[property] != tool.image.properties[property]) {
				return false;
			}
		}
		for (const animationProperty in animation) {
			if (animation[animationProperty] != tool.image.animation[animationProperty]) {
				return false;
			}
		}
		return true;
	}

	clone() {
		const { name, type, width, height, layer, color, amount, taper, image, capacity, intervals } = this;
		const newPosition = this.position.clone();
		const newImage = this.cloneImage(type, image);
		const clonedTool = new Tool(
			type,
			newImage,
			newPosition,
			width,
			height,
			layer,
			color,
			amount,
			taper,
			capacity,
			intervals
		);
		clonedTool.setName(name);
		return clonedTool;
	}

	cloneImage(type, image) {
		const clonedImage = createImage(type);
		const { properties, animation } = image;
		for (const property in properties) {
			clonedImage.properties[property] = image.properties[property];
		}
		if (animation) {
			for (const animation_property in animation) {
				clonedImage.animation[animation_property] = image.animation[animation_property];
			}
		}
		return clonedImage;
	}

	resize(widthRatio, heightRatio) {
		const { x, y } = this.position;
		const { width, height } = this;
		const scaledX = x * widthRatio;
		const scaledY = y * heightRatio;
		const scaledWidth = width * widthRatio;
		const scaledHeight = height * heightRatio;

		this.position = new Position(scaledX, scaledY);
		this.width = scaledWidth;
		this.height = scaledHeight;
	}

	toString() {
		return this.name;
	}

	toSelectOption() {
		return {
			label: this.name,
			value: this
		};
	}

	static load(toolData) {
		const {
			toolType,
			x,
			y,
			width,
			height,
			toolIdentity,
			color,
			amount,
			capacity,
			intervals,
			taper,
			name
		} = toolData;
		const { layer } = toolIdentity;
		const image = Tool.loadImage(toolType, { color, amount, capacity, intervals, taper });
		const position = new Position(x, y);
		const loadedTool = new Tool(
			toolType,
			image,
			position,
			width,
			height,
			layer,
			color,
			amount,
			taper,
			capacity,
			intervals
		);
		loadedTool.setName(name);
		return loadedTool;
	}

	static loadImage(toolType, properties) {
		const { color, amount, capacity, intervals, taper } = properties;
		const image = createImage(toolType);
		if (image.properties.hasOwnProperty('Fill')) {
			image.properties.Fill = amount;
			image.properties.Color = color;
		}
		if (image.properties.hasOwnProperty('Capacity')) {
			image.properties.Capacity = capacity;
			image.properties.Intervals = intervals;
		}
		if (image.properties.hasOwnProperty('Taper')) {
			image.properties.Taper = taper;
		}
		return image;
	}

	save(lessonId, stepNumber) {
		return {
			toolIdentity: {
				layer: this.layer,
				stepNumber: stepNumber,
				lessonId: lessonId
			},
			toolType: this.type,
			x: this.position.x,
			y: this.position.y,
			name: this.name,
			amount: this.amount,
			color: this.color,
			height: this.height,
			width: this.width,
			taper: this.taper,
			capacity: this.capacity,
			intervals: this.intervals
		};
	}
}

export default Tool;

File: backend/Bru/src/main/java/BruGroup/Bru/Entity/Tool.java
package BruGroup.Bru.Entity;
import javax.persistence.*;
import javax.validation.constraints.NotNull;

@Table(name = "Tool")
@Entity
public class Tool {

    @EmbeddedId
    private ToolIdentity toolIdentity;

    @NotNull
    private String toolType;

    @NotNull
    private double x;

    @NotNull
    private double y;

    private String name;
    private double amount;
    private String color;
    private double height;
    private double width;
    private double taper;
    private double capacity;
    private double intervals;

    public Tool() {

    }

    public Tool(ToolIdentity toolIdentity,
                String toolType,
                double x,
                double y,
                double amount,
                String color,
                String name,
                double width,
                double height,
                double taper,
                double capacity,
                double intervals) {
        this.toolIdentity = toolIdentity;
        this.toolType = toolType;
        this.x = x;
        this.y = y;
        this.name = name;
        this.amount = amount;
        this.color = color;
        this.height = height;
        this.width = width;
        this.taper = taper;
        this.capacity = capacity;
        this.intervals = intervals;

    }

    public Tool clone(int lessonId) {
        ToolIdentity clonedIdentity = this.toolIdentity.clone(lessonId);
        return new Tool(
                clonedIdentity,
                this.toolType,
                this.x,
                this.y,
                this.amount,
                this.color,
                this.name,
                this.width,
                this.height,
                this.taper,
                this.capacity,
                this.intervals
        );
    }

    public ToolIdentity getToolIdentity() {
        return toolIdentity;
    }

    public void setToolIdentity(ToolIdentity toolIdentity) {
        this.toolIdentity = toolIdentity;
    }

    public String getToolType() {
        return toolType;
    }

    public void setToolType(String toolType) {
        this.toolType = toolType;
    }

    public double getX() {
        return x;
    }

    public void setX(double x) {
        this.x = x;
    }

    public double getY() {
        return y;
    }

    public void setY(double y) {
        this.y = y;
    }

    public double getAmount() {
        return amount;
    }

    public void setAmount(double amount) {
        this.amount = amount;
    }

    public String getColor() {
        return color;
    }

    public void setColor(String color) {
        this.color = color;
    }

    public double getHeight() {return height;}

    public double getWidth() {return width;}

    public void setHeight(double height) {this.height = height;}

    public void setWidth(double width) {this.width = width;}

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public double getTaper() {
        return taper;
    }

    public void setTaper(double taper) {
        this.taper = taper;
    }

    public double getCapacity() {
        return capacity;
    }

    public void setCapacity(double capacity) {
        this.capacity = capacity;
    }

    public double getIntervals() {
        return intervals;
    }

    public void setIntervals(double intervals) {
        this.intervals = intervals;
    }
}
